pkg_resources/__init__.py

```diff
@@ -99,7 +101,16 @@
 from pkg_resources.extern.platformdirs import user_cache_dir as _user_cache_dir
 
 if TYPE_CHECKING:
-    from _typeshed import StrPath
+    from typing_extensions import Self
+    from _typeshed import StrPath, StrOrBytesPath
+    from itertools import chain
+
+    # yield_lines is exported, we should type it, but extern is too dynamic
+    def yield_lines(iterable: "_NestedStr") -> chain[str]: ...
+
+    # Trick type-checkers into seeing the original instances for type comparisons
+    from packaging import version as _packaging_version
+    from packaging import requirements as _packaging_requirements
```

Comment:
I think we should defer any changes like this until we have a decision about how to reconcile `extern` and type-cheking.

```python
if TYPE_CHECKING:
    # ...
    def yield_lines(iterable: "_NestedStr") -> chain[str]: ...

    # Trick type-checkers into seeing the original instances for type comparisons
    from packaging import version as _packaging_version
    from packaging import requirements as _packaging_requirements
```

Reply:
Even if we did, this part would be useful. At least the two imports.

Even if static type-checkers understood the `extern` imports, they would be treated as a different type than the originals:

```py
from pkg_resources.parse_version
from packaging import Version

def foo(bar: Version): ...
foo(parse_version()):... # type error
```
(I'm on my phone typing from memory, will update this comment with actual type issue soon).

The re-exported `yield_lines` method isn't typed either iirc. (Although that can be fixed upstream)

Or would you prefer keeping even that for later, adding on top of the `extra` discussion ?